### PR TITLE
Avoid `alloca`s in codegen for simple `mir::Aggregate` statements

### PR DESCRIPTION
**Pull request by scottmcm**
The core idea here is to remove the abstraction penalty of simple newtypes in codegen.

Even something simple like constructing a
```rust
#[repr(transparent)] struct Foo(u32);
```
forces an `alloca` to be generated in nightly right now.

Certainly LLVM can optimize that away, but it would be nice if it didn't have to.

Quick example:
```rust
#[repr(transparent)]
pub struct Transparent32(u32);

#[no_mangle]
pub fn make_transparent(x: u32) -> Transparent32 {
    let a = Transparent32(x);
    a
}
```
on nightly we produce <https://rust.godbolt.org/z/zcvoM79ae>
```llvm
define noundef i32 @make_transparent(i32 noundef %x) unnamed_addr #0 {
  %a = alloca i32, align 4
  store i32 %x, ptr %a, align 4
  %0 = load i32, ptr %a, align 4, !noundef !3
  ret i32 %0
}
```
but after this PR we produce
```llvm
define noundef i32 @make_transparent(i32 noundef %x) unnamed_addr #0 {
start:
  ret i32 %x
}
```
(even before the optimizer runs).

### CODE REVIEW

**Comment by fee1-dead** [comment]
r? compiler

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs [hunk 1]
```diff
@@ -720,12 +720,44 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 OperandRef { val: OperandValue::Immediate(static_), layout }
             }
             mir::Rvalue::Use(ref operand) => self.codegen_operand(bx, operand),
-            mir::Rvalue::Repeat(..) | mir::Rvalue::Aggregate(..) => {
+            mir::Rvalue::Repeat(..) => {
                 // According to `rvalue_creates_operand`, only ZST
-                // aggregate rvalues are allowed to be operands.
+                // repat rvalues are allowed to be operands.
```
Comment (klensy) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:725]: repeat

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs [hunk 2]
```diff
@@ -720,12 +720,44 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
                 let ty = rvalue.ty(self.mir, self.cx.tcx());
                 OperandRef::zero_sized(self.cx.layout_of(self.monomorphize(ty)))
             }
+            mir::Rvalue::Aggregate(ref kind, ref fields) => {
+                let ty = rvalue.ty(self.mir, self.cx.tcx());
+                let ty = self.monomorphize(ty);
+                let layout = self.cx.layout_of(self.monomorphize(ty));
```
Comment (klensy) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:732]: Is second call to `monomorphize` required?

**Comment by lqd** [comment]
(I've seen regex and wg-grammar being noisy in other PRs)

File: compiler/rustc_codegen_ssa/Cargo.toml [hunk 3]
```diff
@@ -6,8 +6,10 @@ edition = "2021"
 [dependencies]
 # tidy-alphabetical-start
 ar_archive_writer = "0.1.5"
+arrayvec = { version = "0.7", default-features = false }
 bitflags = "2.4.1"
 cc = "1.0.90"
+either = "1.5.0"
 itertools = "0.12"
```
[the remark this answers is not in the document]
Comment (mati865) [reply compiler/rustc_codegen_ssa/Cargo.toml:13]: Have workspace dependencies been considered in this repo?
It'd make it much easier to keep versions in sync.

[the remark this answers is not in the document]
Comment (klensy) [reply compiler/rustc_codegen_ssa/Cargo.toml:13]: No version dupes in cargo.lock, so all ok.

[the remark this answers is not in the document]
Comment (mati865) [reply compiler/rustc_codegen_ssa/Cargo.toml:13]: The fact there are no duplicates doesn't mean it's all ok. Having those versions in a single place would prevent the problem from appearing in the first place.

**Comment by jieyouxu** [comment]
r? compiler (this touches mir which I am not familiar with :3)

**Comment by jieyouxu** [comment]
r? oli-obk (feel free to reassign, didn't want to randomly roll)

**Comment by jieyouxu** [comment]
Oh wait oli is on vacation 🤦
r? mir

**Comment by nnethercote** [comment]
Nice binary size reductions. They are almost entirely in debug builds, indicating that "Certainly LLVM can optimize that away" is correct but that only happens in opt builds :)

File: tests/codegen/mir-aggregate-no-alloca.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+// CHECK-LABEL: i8 @make_cell_of_bool(i1 noundef zeroext %b)
+#[no_mangle]
+pub fn make_cell_of_bool(b: bool) -> std::cell::Cell<bool> {
+    // CHECK: %[[BYTE:.+]] = zext i1 %b to i8
+    // CHECK: ret i8 %[[BYTE]]
+    std::cell::Cell::new(b)
+}
+
+// CHECK-LABLE: { i8, i16 } @make_cell_of_bool_and_short(i1 noundef zeroext %b, i16 noundef %s)
```
Comment (aapoalas) [inline tests/codegen/mir-aggregate-no-alloca.rs:58]: Typo here: "CHECK LABLE" should be "LABEL". The test is presumably not working properly.

**Comment by Kobzol** [comment]
Nice binary size reductions. The regressions are limited to a single benchmark (`bitmaps`), while the wins are across three different (primary) benchmarks.

@rustbot label: +perf-regression-triaged

**Pull request merged**